salesforce_functions/data_api/exceptions.py

```diff
@@ -1,2 +1,31 @@
+from dataclasses import dataclass
+
+
+@dataclass(frozen=True, slots=True)
+class InnerSalesforceRestApiError:
+    """An error returned from the Salesforce REST API."""
+
+    message: str
+    error_code: str
+    fields: list[str]
+
+
 class DataApiError(Exception):
-    pass
+    """Base class for Data API exceptions"""
+
+
+@dataclass(frozen=True, slots=True)
```

Comment:
This is me wanting to avoid `__repr__` for the ones that have fields. That said, I'm actually not 💯 sure which of these `__` methods you are supposed to implement for exceptions. Absolutely no pushback from me changing this. :)

Reply:
So for a typical custom exception that adds attributes, one overrides `__init__` in order to set the custom attributes, and the `__init__` should also call the base `Exception` class's `__init__`, so that some of the underlying exception implementation details work.

One query I had about using a dataclass, was whether the automagically created class overrode the base class's `__init__` and so break the above behaviour?

Reply:
Hmm though using GitHub codesearch (thanks to the new beta's regex support 🎉 ) I can see there are a couple of projects using dataclasses with exceptions, so maybe it's fine? 
https://github.com/search?q=%2Fdataclass.*%5Cnclass+.%2BException%5C%28%2F+language%3APython&type=code&l=Python

As such, let's maybe leave as-is for now.